.ci/magician/cmd/test_eap_vcr.go

```diff
@@ -0,0 +1,250 @@
+		if err != nil {
+			return fmt.Errorf("error formatting action taken comment: %w", err)
+		}
+		comment := strings.Join([]string{testsAnalyticsComment, withReplayFailedTestsComment}, "\n")
+		if err := postGerritComment(kokoroArtifactsDir, modifiedFilePath, comment, rnr); err != nil {
+			return fmt.Errorf("error posting comment: %w", err)
+		}
+
+		recordingResult, recordingErr := vt.RunParallel(vcr.RunOptions{
+			Mode:     vcr.Recording,
+			Version:  provider.Beta,
+			TestDirs: testDirs,
+			Tests:    replayingResult.FailedTests,
+		})
+
+		if hasPanics, err := handleEAPVCRPanics(head, kokoroArtifactsDir, modifiedFilePath, recordingResult, vcr.Recording, rnr); err != nil {
+			return fmt.Errorf("error handling panics: %w", err)
+		} else if hasPanics {
+			return nil
+		}
+		replayingAfterRecordingResult := vcr.Result{}
+		if len(recordingResult.PassedTests) > 0 {
+			replayingAfterRecordingResult, _ = vt.RunParallel(vcr.RunOptions{
+				Mode:     vcr.Replaying,
+				Version:  provider.Private,
+				TestDirs: testDirs,
+				Tests:    recordingResult.PassedTests,
+			})
+			if err := vt.UploadLogs(vcr.UploadLogsOptions{
+				Head:           head,
+				Parallel:       true,
+				AfterRecording: true,
+				Mode:           vcr.Recording,
+				Version:        provider.Private,
+			}); err != nil {
+				return fmt.Errorf("error uploading recording logs: %w", err)
+			}
+		}
+		hasTerminatedTests := (len(recordingResult.PassedTests) + len(recordingResult.FailedTests)) < len(replayingResult.FailedTests)
+		allRecordingPassed := len(recordingResult.FailedTests) == 0 && !hasTerminatedTests && recordingErr == nil
+		recordReplayData := recordReplay{
+			RecordingResult:               recordingResult,
+			ReplayingAfterRecordingResult: replayingAfterRecordingResult,
+			RecordingErr:                  recordingErr,
+			HasTerminatedTests:            hasTerminatedTests,
+			AllRecordingPassed:            allRecordingPassed,
+		}
+		recordReplayComment, err := formatRecordReplay(recordReplayData)
+		if err != nil {
+			return fmt.Errorf("error formatting record replay comment: %w", err)
+		}
+		if err := postGerritComment(kokoroArtifactsDir, modifiedFilePath, recordReplayComment, rnr); err != nil {
+			return fmt.Errorf("error posting comment: %w", err)
+		}
+	} else { //  len(replayingResult.FailedTests) == 0
+		withoutReplayFailedTestsData := withoutReplayFailedTests{
+			ReplayingErr: replayingErr,
+		}
+		withoutReplayFailedTestsComment, err := formatWithoutReplayFailedTests(withoutReplayFailedTestsData)
+		if err != nil {
+			return fmt.Errorf("error formatting action taken comment: %w", err)
+		}
+		comment := strings.Join([]string{testsAnalyticsComment, withoutReplayFailedTestsComment}, "\n")
+		if err := postGerritComment(kokoroArtifactsDir, modifiedFilePath, comment, rnr); err != nil {
+			return fmt.Errorf("error posting comment: %w", err)
+		}
+	}
+	return nil
+}
+
+var allowedAlphaServices = map[string]struct{}{
```

Comment:
How/why were these services chosen? Also, we may actually want a disallowed list, so that we don't need to update this every time a new service is added to EAP.

Reply:
These are the services that have private overrides.

My plan was to remove this restriction altogether once the EAP environment is fully set up. I could also make a disallowed list, but it would be much longer.

Reply:
Yea, a significant percentage of the contributions in EAP use new services, so I think we need to support testing for those out-of-the-box. I would actually tweak this to be a disallow list for the services that we know or suspect have problems due to the test environment not being set up. But even then, I would be curious if we could just omit this entirely and let the tests run without the env vars, and that way when we update the test environment we don't need to make additional changes.